src/main/java/io/nats/client/ConnectionImpl.java

```diff
@@ -1811,18 +1827,27 @@ SubscriptionImpl subscribe(String subject, String queue, MessageHandler cb,
             }
 
             if (cb != null) {
-                sub = new AsyncSubscriptionImpl(this, subject, queue, cb);
-                // If we have an async callback, start up a sub specific Runnable to deliver the
-                // messages
-                subexec.submit(new Runnable() {
-                    public void run() {
-                        try {
-                            waitForMsgs((AsyncSubscriptionImpl) sub);
-                        } catch (InterruptedException e) {
-                            Thread.currentThread().interrupt();
+                // Check if we should use the global message delivery pool or create our own thread.
+                MsgDeliveryPool msgDlvPool = null;
+                final boolean   useDlvPool = (this.opts.useGlobalMsgDelivery &&
+                                                ((msgDlvPool = Nats.getMsgDeliveryThreadPool()) != null));
+
+                sub = new AsyncSubscriptionImpl(this, subject, queue, cb, useDlvPool);
+                if (useDlvPool) {
+                    msgDlvPool.assignDeliveryWorker((AsyncSubscriptionImpl) sub);
+                } else {
+                    // If we have an async callback, start up a sub specific Runnable to deliver the
+                    // messages
+                    subexec.submit(new Runnable() {
+                        public void run() {
+                            try {
+                                waitForMsgs((AsyncSubscriptionImpl) sub);
+                            } catch (InterruptedException e) {
+                                Thread.currentThread().interrupt();
```

Comment:
Should we do anything besides setting the interrupt flag? Is this handled elsewhere?

Reply:
I changed `waitForMsgs()` to catch any throwable that the user callback should issue. There is no need for this try/catch here, will remove.